site/posts/python-tutorial-faq-part-2.ipynb

```diff
@@ -0,0 +1,153 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here is a compilation of questions and issues that arose during the second session of the Python Tutorial Seminar Series, on creating a data dictionary. After this first short series, these FAQ sections will be added to the [Xdev tutorial website](https://ncar.github.io/python-tutorial/)\n",
+    "\n",
+    "------"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Q.** Why is `list[0:10]` the first 10 elements, not 11?\n",
+    "\n",
+    "**A.** In Python the second number is exclusive so you read up to (but excluding) that index. \n",
+    "\n",
+    "------\n",
+    "<!-- TEASER_END -->"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Q.** Why do we not say `list[1,10]` to returnthe first 10 elements?\n",
+    "\n",
+    "**A.** Python is 0-indexed, so index of 1 actually refers to the second element. Say our list was `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]`, then `list[1,10]` would return `[1, 2, 3, 4, 5, 6, 7, 8, 9]` (which is missing the first element) and `list[0,10]` returns the desired `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`.\n",
+    "\n",
+    "------"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Q.** So `list[0:1]` is just one value, same as `list[0]`? \n",
+    "\n",
+    "**A.** Actually not. `list[0:1]` returns a list, due to the slice indexing, of value `[0]`, whereas `list[0]` returns the 0th element, and has value `0`.\n",
+    "\n",
+    "------"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Q.** How do you read only the 1st, 4rd, 7th, 10th, . . . rows of a data?\n",
+    "\n",
+    "**A.** Using `list[::3]`\n",
+    "\n",
+    "------"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Q.** This part is not still clear:\n",
+    "```\n",
+    "# Read the first three lines (header)\n",
+    "for _ in range(3):\n",
+    "    datafile.readline()\n",
+    "```\n",
+    "Could you clarify it?\n",
+    "\n",
+    "**A.** `range(3)` is an iterator; it is equivalent to saying `for _ in [0,1,2]` or just any list with length 3. Here we’re iterating over that 3-element list and using the value for each iteration as the value of the variable `_`. We need to define the variable `_` in order to iterate through the for-loop, but we never call it. In every iteration, we call up `datafile.readline()` which means we read 3 lines, but since we aren’t saving them anywhere we are only reading and discarding them. Anything that begins with an underscore is generally considered a discardable by common practice. This for-loop sets us up so the next readline is where the data values begin.\n",
```

Comment:
```suggestion
    "**A.** `range(3)` is an iterator; it is equivalent to using the list `[0,1,2]`.  Here we’re iterating over that 3-element list and using the value for each iteration as the value of the variable `_`.  (We need to define the variable `_` in order to iterate through the for-loop, but we never use it, which is why we just call it `_`.).  In every iteration, we call `datafile.readline()` which means we read 3 lines of text from the file *in total*, but since we aren’t saving those lines of text anywhere (i.e., in any variable) we are only reading and discarding them. (Note, anything that begins with an underscore is generally considered "discardable" by common practice, but there is no rule in Python about that.) This for-loop sets us up so the next `readline()` call is where the data values begin.\n",
```